### PR TITLE
Topic/monitoring

### PR DESCRIPTION
**Pull request by bosilca**
Add a new capability to Open MPI. With this patch, OMPI gain the capability to monitor all communications and to dump the communication heat map (including differences between internal and user communications) for point-to-point, collective and RMA. The entire framework is driven using the new, MPI Forum blessed, MPI_T interface.

Fully fledged documentation and scripts to manipulate the generated data are also provided.

### CODE REVIEW

File: ompi/mca/pml/monitoring/README [hunk 1]
```diff
@@ -20,6 +21,17 @@ communication channels. At that levels all communication are point-to-point comm
...
+Low level communication monitoring interface in OpenMPI
+
+Introduction
+------------
+This inteface traces and monitors all messages sent by MPI before they go to the
+communication channels. At that levels all communication are point-to-point communications:
+collectives are already decomposed in send and receive calls.
+
+The monitoring is strored internally by each process and output on stderr at the end of the
+>>>>>>> Add a monitoring PML. This PML track all data exchanges by the processes
```
Comment (miked-mellanox) [inline ompi/mca/pml/monitoring/README]: merge leftovers

Comment (bosilca) [inline ompi/mca/pml/monitoring/README]: thanks. fixed.

**Comment by hjelmn** [comment]
:bot:retest:

**Comment by hppritcha** [comment]
lanl distcheck doesn't like this
```
make[3]: Entering directory `/home/hppritcha/jenkins/workspace/ompi_master_pr_distcheck/openmpi-gitclone/_build/ompi/mca/common/monitoring'
  CC       libmca_common_monitoring_la-common_monitoring.lo
  CC       libmca_common_monitoring_la-common_monitoring_coll.lo
  LN_S     libmca_common_monitoring.la
../../../../../ompi/mca/common/monitoring/common_monitoring.c:17:31: fatal error: common_monitoring.h: No such file or directory
 #include "common_monitoring.h"
                               ^
compilation terminated.
../../../../../ompi/mca/common/monitoring/common_monitoring_coll.c:17:31: fatal error: common_monitoring.h: No such file or directory
 #include "common_monitoring.h"
                               ^
compilation terminated.
```

**Comment by hppritcha** [comment]
botany bay doesn't like this either:
```
  CC       monitoring_prof.lo
monitoring_prof.c:385:6: warning: no previous prototype for function 'monitoring_prof_mpi_init_f2c' [-Wmissing-prototypes]
void monitoring_prof_mpi_init_f2c( MPI_Fint *ierr ) { 
     ^
monitoring_prof.c:394:6: warning: no previous prototype for function 'monitoring_prof_mpi_finalize_f2c' [-Wmissing-prototypes]
void monitoring_prof_mpi_finalize_f2c( MPI_Fint *ierr ) { 
     ^
In file included from monitoring_prof.c:417:
../../ompi/mpi/fortran/mpif-h/bindings.h:67:2: error: Unrecognized Fortran name mangling scheme
#error Unrecognized Fortran name mangling scheme
 ^
monitoring_prof.c:424:28: error: expected identifier
                           (MPI_Fint *ierr),
                           ^
monitoring_prof.c:427:1: error: expected function body after function declarator
OMPI_GENERATE_F77_BINDINGS (MPI_FINALIZE,
^
2 warnings and 3 errors generated.
make[2]: *** [monitoring_prof.lo] Error 1
make[1]: *** [install-recursive] Error 1
make: *** [install-recursive] Error 1
Build step 'Execute shell' marked build as failure
```

**Comment by hppritcha** [comment]
the sign off checker doesn't like a lot of these commits either.  
Please squash down the commits in this PR at least some.

**Comment by bosilca** [comment]
I don't see how it cannot find the "common_monitoring.h" file. We are supposed to automatically include "." so the file should be readily available.

I am really sorry for the sign-off checker. The commits will be squashed anyway once merged.

**Comment by ggouaillardet** [comment]
@bosilca that was just a typo that causes ```common_monitoring.h``` is missing from the tarball
an other error is related to lack of weak symbols and no fortran bindings

these should be both fixes in #3131
once it passes CI, i will PR vs your branch so you can simply merge it

**Comment by bosilca** [comment]
Thanks @ggouaillardet, I haven't noticed the failure was on due "make dist".

**Comment by ggouaillardet** [comment]
@bosilca i filed bosilca/ompi#2 to fix all the glitches with this PR

**Comment by bosilca** [comment]
Thanks @ggouaillardet, I merged your PR.

**Comment by ggouaillardet** [comment]
@bosilca i filed bosilca/ompi#3 to silence some misc warnings

**Comment by hppritcha** [comment]
This missed the whitelist for 3.0.0.

**Comment by bosilca** [comment]
Troubling.

**Comment by clementFoyer** [comment]
Hi everyone,
Is there a reason to have the monitored values typed as 64 bits integers instead of 64 bits unsigned integers ? I asked that as the performance variables are defined as MCA_BASE_VAR_TYPE_UNSIGNED_LONG, and to check wether the collective monitoring API as to be adapted as well.

**Comment by ggouaillardet** [comment]
@clementFoyer iirc, i moved this to signed integer in order to silence warnings generated by ```opal_atomic_add()``` and friends.

**Comment by bosilca** [comment]
@ggouaillardet patch reduced the range of the monitoring variables, but I figure it out that as the operations themselves would have had the same effect this should not create additional problems.

To be correct you will have to create your type, because int64_t bits might be either long or long long depending on the platforms.

**Comment by clementFoyer** [comment]
So maybe everything should be changed to size_t variables, and we adapt the MCA_BASE_VAR_TYPE_  and the opal_tomic_add depending to the config.h.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/62b69619479fdb16dbba2d67a4cce002

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/8ce096c87a8e65c024682e560317153b

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3b393f2c564017f1a7c89661cec05835

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/034f66542af6a2ed4d6b188acd85a9fb

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/292eaec3fcbfd9e207168a090e70d8a1

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5722d4527f3a2834f820205ff878ae28

**Comment by clementFoyer** [comment]
It seems like the IBM CI does not generate the OPAL_PTRDIFF_TYPE symbol in opal_config.h (included in ompi_config.h, included in my header files). Is it possible to check the generated files ?

**Comment by bosilca** [comment]
Due to the requirement of C99, we are not using OPAL_PTRDIFF_TYPE anymore.
Use directly ptrdiff_t instead.

  George.

**Comment by clementFoyer** [comment]
I saw the changes in the trunk. I adapted the code accordingly ;) Thank's George !

**Comment by bosilca** [comment]
As far as we are concerned this PR is ready. I plan to merge Friday EOB.

**Comment by clementFoyer** [comment]
bot:retest

**Comment by clementFoyer** [comment]
:bot:retest:

**Pull request merged**